src/conduits/RwaOutputConduit3.sol

```diff
@@ -0,0 +1,325 @@
+// SPDX-FileCopyrightText: © 2020-2021 Lev Livnev <lev@liv.nev.org.uk>
+// SPDX-FileCopyrightText: © 2022 Dai Foundation <www.daifoundation.org>
+// SPDX-License-Identifier: AGPL-3.0-or-later
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Affero General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with this program.  If not, see <https://www.gnu.org/licenses/>.
+pragma solidity 0.6.12;
+
+import {DaiAbstract} from "dss-interfaces/dss/DaiAbstract.sol";
+import {PsmAbstract} from "dss-interfaces/dss/PsmAbstract.sol";
+import {GemAbstract} from "dss-interfaces/ERC/GemAbstract.sol";
+import {GemJoinAbstract} from "dss-interfaces/dss/GemJoinAbstract.sol";
+
+/**
+ * @author Lev Livnev <lev@liv.nev.org.uk>
+ * @author Nazar Duchak <nazar@clio.finance>
+ * @title An Output Conduit for real-world assets (RWA).
+ * @dev This contract differs from the original [RwaOutputConduit](https://github.com/makerdao/MIP21-RWA-Example/blob/fce06885ff89d10bf630710d4f6089c5bba94b4d/src/RwaConduit.sol#L41-L118):
+ *  - The caller of `push()` is not required to hold MakerDAO governance tokens.
+ *  - The `push()` method is permissioned.
+ *  - `push()` permissions are managed by `mate()`/`hate()` methods.
+ *  - `pick` whitelist is managed by `kiss() / diss()` methods.
+ *  - Requires a PSM address in the constructor.
+ *  - `pick` can be called to set the `to` address. Eligible `to` addresses should be whitelisted by an admin through `kiss`.
+ *  - The `push()` method swaps DAI to GEM using PSM and set `to` to zero address.
+ *  - The `push()` method with `wad` argument swaps specified amount of DAI to GEM using PSM and set `to` to zero address.
+ */
+contract RwaOutputConduit3 {
+    /// @notice PSM GEM token contract address.
+    GemAbstract public immutable gem;
+    /// @notice PSM contract address.
+    PsmAbstract public immutable psm;
+    /// @dev DAI/GEM resolution difference.
+    uint256 private immutable to18ConvertionFactor;
+
+    /// @notice Addresses with admin access on this contract. `wards[usr]`
+    mapping(address => uint256) public wards;
+    /// @dev Addresses with operator access on this contract. `can[usr]`
+    mapping(address => uint256) public can;
+
+    /// @dev Not used
+    address private __unused_gov;
+    /// @notice Dai token contract address.
+    DaiAbstract public dai;
+    /// @notice Dai Recipient address.
+    address public to;
```

Comment:
@godsflaw Would you be willing to drop the opinion of keeping the same storage slots? I really think for this case just makes the code uglier without a strong reason.
I'd propose here the following:
- Make `dai` immutable and move it above
- remove `__unused_gov`
- Move `to` before or after the mappings (so the 4 of them remain grouped)

cc @amusingaxl

Reply:
Actually move `to` after the mappings, we want `wards` on storage 0.

Reply:
I left the opinion in a comment above.

Reply:
Thanks, same here.

Reply:
resolved now.